### DISCUSSION TITLE
How define global fetch hook

### DISCUSSION THREAD

**Discussion by silencerspirit** [nuxt/nuxt#17454] [Questions] [answered]
Hi there!
I haven't found any correct implementation for global hooks nuxt (useasyncdata, $fetch, usefetch).
I would like to be able to configure a global hook that will have an Authorization header, base Url, proxy, etc.

Is there any good architectural solution to this issue?

**Comment by danielroe** [comment, marked as answer]
Yes, you should do this with your own composable, e.g. `~/composables/useMyFetch.ts`:

```js
export const useMyFetch = (url) => useFetch(url, { /* your options */ })
```

**Reply by zipme** [reply]
Hi @danielroe , in case we use both useFetch (and useLazyFetch) and useAsyncData (where we then use `$fetch` internally), does that mean we need to create a wrapper composable for all these? Would it be possible to define the global $fetch ourselves?

**Reply by danielroe** [reply]
If you want a global, then you should create your own and not redefine `$fetch`, which would have ... unpredictable ... results.

**Reply by ilyaDegtyarenko** [reply]
When I use another composable function in the **onRequest** option, the request is repeated on the client **when the page refreshes**. That is, the request is executed twice, on the server, then on the client.

If I use native **useFetch** and add a header for it in the page component, then the client has no second request, everything is fine. What could be the reason?

_**useMyFetch composable**_
```
import {useFetch, useCookie} from '#app'

export const useMyFetch = (request) => useFetch(request, {
    async onRequest({options}) {
        options.headers = {
            '[name]': `value - ${useCookie('[name]').value}`
        }
    }
})
```

**_page_**
```
<template>
    <div>
        {{ data }}
    </div>
</template>

<script
    lang="ts"
    setup
>
    import {useMyFetch} from '~/composables/useMyFetch'

    const {data} = await useMyFetch(`[url]`)
</script>
```
**_network in browser_**

<img width="580" alt="Знімок екрана 2022-06-11 о 02 47 36" src="https://user-images.githubusercontent.com/37239202/173163431-fcf18313-6094-4362-b6a7-6dfc1187361d.png">


P.S. Removed all layouts, plugins and middleware for clean testing. Only page.

**Reply by danielroe** [reply]
I'll happily take a look at a reproduction.

**Reply by ilyaDegtyarenko** [reply]
@danielroe here please

https://stackblitz.com/edit/nuxt-starter-acjg3u?file=composables/useMyFetch.ts

**Reply by danielroe** [reply]
The issue is that we need an explicit key if you provide any functions as arguments in the options to `useFetch`. So:

```diff
  import {useFetch, useCookie} from '#app'
  
  export const useMyFetch = (request) => useFetch(request, {
+   key: request,
    async onRequest({options}) {
      // ...
    }
  })
```

**Reply by ilyaDegtyarenko** [reply]
Thank you 😊

**Reply by ilyaDegtyarenko** [reply]
@danielroe Faced with an implicit situation in such an implementation. Needs to be careful with setting the **key** with a static url, because if **initialCache** is enabled, the request will not be sent again when form is changed, because the **key** is not changed.
